### DISCUSSION TITLE
How to define the new filter arguments in graphql

### DISCUSSION THREAD

**Discussion by jwld** [prisma/prisma#2723] [Q&A] [answered]
We're not using nexus and have defined our own graphql layer instead. We're currently set up so that our queries correspond directly with prisma's filtering, so that we can query our DB using prisma's syntax from the front-end.

For example given this schema:
```
model Booking {
  approved: Boolean
}
```

We have the following where input type for prisma 1:
```
input BookingWhereInput {
  approved: Boolean
  approved_not: Boolean
}
```

However I'm totally stumped on how to implement prisma 2's filtering in graphql, since now approved can be either a boolean, or the full filter object (with `equals`, `not` etc.), or even just `null`. Is it possible to define a graphql input type that matches the prisma 2 filters?

Something like this, although I know this isn't possible:
```
input BooleanWhereInput {
  equals: Boolean
  not: Boolean
}

input BookingWhereInput {
  approved: Boolean | BooleanWhereInput | null
}
```

**Comment by ryands17** [comment]
Hey @jwld :wave: 
You can do it in the following manner:

As an example I am adding where to the query 

```graphql
type Query {
  users(where: UserWhereInput): [User!]!
}

type User {
  createdAt: DateTime!
  isAdmin: Boolean
  name: String
}

input UserWhereInput {
  email: StringFilter
  id: IntFilter
  isAdmin: BooleanFilter
  name: NullableStringFilter
}

input BooleanFilter {
  equals: Boolean
  not: Boolean
}
```

This is how a `BooleanFilter` would be created. You can create this by looking at the generated Prisma types as well as that would help.

Let me know if this works for you.

**Reply by jwld** [reply]
Thanks - the thing I'm stuck on is that with prisma 2 it can be filtered in various ways - either using the object (equals, not, contains etc.), or by using a string (effectively a shortcut for "equals").
```
where: { 
  name: "Harry"
}
```
OR
```
where: {
  name: {
    equals: "Harry"
  }
}
```
OR
```
where: {
  name: null
}
```

Is it possible to replicate this behaviour with graphql input types?

**Reply by ryands17** [reply]
Yes youl would have to create a `StringFilter` in the same manner:

```graphql
input StringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: String
  notIn: [String!]
  startsWith: String
}
```
You can use equals and other operators, but cannot pass the value directly.

**Comment by jwld** [comment, marked as answer]
Sorry yea I understand that - what I'm specifically asking is whether it's possible to set up our graphql layer so that I can query it from the front-end as _either_ a string, _or_ using the more complex object (equals, not etc.). I understand how to define the filter types.

I want to be able to do both of the following from the front-end (i.e. match prisma's query filtering behaviour exactly):
```
where({ name: "Foo" })
```
```
where({ name: { not: "Foo" } })
```

**Reply by ryands17** [reply]
I don't think this is possible an UnionTypes cannot be represented in GraphQL inputs. So either a string or object is niot supported. You can either pass an entire object, or a string.

**Comment by ludralph** [comment]
Hi @jwld,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
